internal/datacoord/channel_store.go

```diff
@@ -0,0 +1,303 @@
+	*cos = append(*cos, &ChannelOp{
+		NodeID:   id,
+		Type:     Delete,
+		Channels: channels,
+	})
+}
+
+// ROChannelStore is the read only channel store from which user can read the mapping between channels and node
+type ROChannelStore interface {
+	// GetNode gets the channel info of node
+	GetNode(nodeID int64) *NodeChannelInfo
+	// GetChannels gets all channel infos
+	GetChannels() []*NodeChannelInfo
+	// GetNodesChannels gets the channels assigned to real nodes
+	GetNodesChannels() []*NodeChannelInfo
+	// GetBufferChannelInfo gets the unassigned channels
+	GetBufferChannelInfo() *NodeChannelInfo
+}
+
+// RWChannelStore is the read write channel store which matains the mapping between channels and node
+type RWChannelStore interface {
+	ROChannelStore
+	// Reload restores the buffer channels and node-channels mapping form kv
+	Reload() error
+	// Add creates a new node-channels mapping, but no channels are assigned to this node
+	Add(nodeID int64)
+	// Delete removes nodeID and returns the channels
+	Delete(nodeID int64) ([]*channel, error)
+	// Update applies the operations in ChannelOpSet
+	Update(op ChannelOpSet) error
+}
+
+// ChannelStore maintains the mapping relationship between channel and datanode
+type ChannelStore struct {
+	store        kv.TxnKV
+	channelsInfo map[int64]*NodeChannelInfo
+}
+
+// NodeChannelInfo is the mapping between channels and node
+type NodeChannelInfo struct {
+	NodeID   int64
+	Channels []*channel
+}
+
+// NewChannelStore creates a new ChannelStore
+func NewChannelStore(kv kv.TxnKV) *ChannelStore {
+	c := &ChannelStore{
+		store:        kv,
+		channelsInfo: make(map[int64]*NodeChannelInfo),
+	}
+	c.channelsInfo[bufferID] = &NodeChannelInfo{
+		NodeID:   bufferID,
+		Channels: make([]*channel, 0),
+	}
+	return c
+}
+
+// Reload restores the buffer channels and node-channels mapping form kv
+func (c *ChannelStore) Reload() error {
+	keys, values, err := c.store.LoadWithPrefix(Params.ChannelWatchSubPath)
+	if err != nil {
+		return err
+	}
+	for i := 0; i < len(keys); i++ {
+		k := keys[i]
+		v := values[i]
+		nodeID, err := parseNodeID(k)
+		if err != nil {
+			return err
+		}
+
+		temp := &datapb.ChannelWatchInfo{}
+		if err := proto.UnmarshalText(v, temp); err != nil {
+			return err
+		}
+
+		c.Add(nodeID)
+		channel := &channel{
+			name:         temp.GetVchan().GetChannelName(),
+			collectionID: temp.GetVchan().GetCollectionID(),
+		}
+		c.channelsInfo[nodeID].Channels = append(c.channelsInfo[nodeID].Channels, channel)
+	}
+	return nil
+}
+
+// Add create a new node-channels mapping, but no channels are assigned to this node
+func (c *ChannelStore) Add(nodeID int64) {
+	if _, ok := c.channelsInfo[nodeID]; ok {
+		return
+	}
+
+	c.channelsInfo[nodeID] = &NodeChannelInfo{
+		NodeID:   nodeID,
+		Channels: make([]*channel, 0),
+	}
+}
+
+// Update applies the operations in opSet
+func (c *ChannelStore) Update(opSet ChannelOpSet) error {
+	for _, v := range opSet {
```

Comment:
how to prevent double assignment?
See if add is received but remove is not received by the old holder, is that possible that data is double consumed and generate two binlogs

Reply:
We check it in SaveBinlogPaths